tests/integration_tests/scheduler/test_lsf_driver.py

```diff
@@ -217,3 +219,60 @@ def mock_poll_once_by_bhist(*args, **kwargs):
     assert "bhist is used" in caplog.text
     assert bhist_called
     assert job_id in driver._bhist_cache
+
+
+async def test_kill_before_submit_is_finished(
+    tmp_path, monkeypatch, caplog, pytestconfig
+):
+    os.chdir(tmp_path)
+
+    if pytestconfig.getoption("lsf"):
+        # Allow more time when tested on a real compute cluster to avoid false positives.
+        job_kill_window = 10
+        test_grace_time = 20
+    elif sys.platform.startswith("darwin"):
+        # Mitigate flakiness on low-power test nodes
+        job_kill_window = 5
+        test_grace_time = 10
+    else:
+        job_kill_window = 1
+        test_grace_time = 2
+
+    bin_path = tmp_path / "bin"
+    bin_path.mkdir()
+    monkeypatch.setenv("PATH", f"{bin_path}:{os.environ['PATH']}")
+    bsub_path = bin_path / "slow_bsub"
+    bsub_path.write_text(
+        "#!/bin/sh\nsleep 0.1\nbsub $@",
+        encoding="utf-8",
+    )
+    bsub_path.chmod(bsub_path.stat().st_mode | stat.S_IEXEC)
+
+    caplog.set_level(logging.DEBUG)
+    driver = LsfDriver(bsub_cmd="slow_bsub")
+
+    # Allow submit and kill to be interleaved by asyncio by issuing
+    # submit() in its own asyncio Task:
+    asyncio.create_task(
+        driver.submit(
+            # The sleep is the time window in which we can kill the job before
+            # the unwanted finish message appears on disk.
+            0,
+            "sh",
+            "-c",
+            f"sleep {job_kill_window}; touch {tmp_path}/survived",
+        )
+    )
+    await asyncio.sleep(0.01)  # Allow submit task to start executing
+    await driver.kill(0)  # This will wait until the submit is done and then kill
+
+    async def finished(iens: int, returncode: int):
+        SIGTERM = 15
+        assert iens == 0
+        assert returncode == SIGNAL_OFFSET + SIGTERM
+
+    await poll(driver, {0}, finished=finished)
+
+    assert "ERROR" not in str(caplog.text)
+    await asyncio.sleep(test_grace_time)
```

Comment:
Why do we need to sleep here?

Reply:
We have in the `poll()` await ensured that the realization has been reported back to the scheduler with  a correct error code. But, if that is erroneous, and the shell script that is running is not killed (it is the mother process of the shell script that is actually killed), then the `survived` file would still appear on disk. So we wait until we know the survived file would have been written and then assert it is not there.

Reply:
Right, but that deserves a comment before the `await asyncio.sleep(test_grace_time)` line.

Reply:
added.